File: lib/utils/classifyPointInBounds.ts
import { pointToBoundsDistance } from "@tscircuit/math-utils"

export type PointBoundsPosition = "on-boundary" | "inside" | "outside"

const isFiniteNumber = ({ value }: { value: number }): boolean =>
  Number.isFinite(value)

const isWithinRange = ({
  value,
  min,
  max,
}: {
  value: number
  min: number
  max: number
}): boolean => value >= min && value <= max

const isWithinEpsilon = ({
  value,
  target,
  epsilon,
}: {
  value: number
  target: number
  epsilon: number
}): boolean => Math.abs(value - target) <= epsilon

export const classifyPointInBounds = ({
  point,
  bounds,
  epsilon = 1e-6,
}: {
  point: { x: number; y: number }
  bounds: { minX: number; maxX: number; minY: number; maxY: number }
  epsilon?: number
}): PointBoundsPosition => {
  const finiteValues = [
    point.x,
    point.y,
    bounds.minX,
    bounds.maxX,
    bounds.minY,
    bounds.maxY,
  ]

  if (!finiteValues.every((value) => isFiniteNumber({ value }))) {
    return "outside"
  }

  const distanceToBounds = pointToBoundsDistance(point, bounds)
  if (distanceToBounds > epsilon) {
    return "outside"
  }

  const isNearMinX = isWithinEpsilon({
    value: point.x,
    target: bounds.minX,
    epsilon,
  })
  const isNearMaxX = isWithinEpsilon({
    value: point.x,
    target: bounds.maxX,
    epsilon,
  })
  const isNearMinY = isWithinEpsilon({
    value: point.y,
    target: bounds.minY,
    epsilon,
  })
  const isNearMaxY = isWithinEpsilon({
    value: point.y,
    target: bounds.maxY,
    epsilon,
  })

  const isNearVerticalEdge = isNearMinX || isNearMaxX
  const isNearHorizontalEdge = isNearMinY || isNearMaxY
  const isNearCorner = isNearVerticalEdge && isNearHorizontalEdge

  if (isNearCorner) {
    return "on-boundary"
  }

  const isWithinYRange = isWithinRange({
    value: point.y,
    min: bounds.minY,
    max: bounds.maxY,
  })
  const isWithinXRange = isWithinRange({
    value: point.x,
    min: bounds.minX,
    max: bounds.maxX,
  })

  const isOnBoundary =
    (isNearVerticalEdge && isWithinYRange) ||
    (isNearHorizontalEdge && isWithinXRange)

  return isOnBoundary ? "on-boundary" : "inside"
}

File: rust/capacity-autorouter/src/bindings/high_density/specialized_utils/classify_point_in_bounds.rs
use super::math::{Bounds, Point};
pub const BOUNDARY_COORDINATE_TOLERANCE_MM: f64 = 1e-6;

pub fn classify_point_in_bounds(point: Point, b: Bounds, epsilon: Option<f64>) -> &'static str {
    let epsilon = epsilon.unwrap_or(BOUNDARY_COORDINATE_TOLERANCE_MM);
    if ![point.x, point.y, b.min_x, b.max_x, b.min_y, b.max_y]
        .iter()
        .all(|v| v.is_finite())
    {
        return "outside";
    }
    if point.x < b.min_x - epsilon
        || point.x > b.max_x + epsilon
        || point.y < b.min_y - epsilon
        || point.y > b.max_y + epsilon
    {
        return "outside";
    }
    let vertical = (point.x - b.min_x).abs() <= epsilon || (point.x - b.max_x).abs() <= epsilon;
    let horizontal = (point.y - b.min_y).abs() <= epsilon || (point.y - b.max_y).abs() <= epsilon;
    if vertical && horizontal {
        return "on-boundary";
    }
    let within_y = point.y >= b.min_y && point.y <= b.max_y;
    let within_x = point.x >= b.min_x && point.x <= b.max_x;
    if vertical && within_y || horizontal && within_x {
        "on-boundary"
    } else {
        "inside"
    }
}
